### PR TITLE
Python unit test updates.

### PR DESCRIPTION
**Pull request by meimchu**
- Rewrote ColorSpaceTest and CDLTransformTest modules.
- Updated OpenColorIOTestSuite module accordingly.
- Added UnitTestUtils module for possible utilities usage.

Signed-off-by: Mei Chu <meimchu@gmail.com>

### CODE REVIEW

**Comment by meimchu** [comment]
The following are errors/warnings raised by these two updated test suites. Based on my current understanding, I believe they should all have passed.

```
test_allocation (ColorSpaceTest.ColorSpaceTest) ... ok
test_allocation_vars (ColorSpaceTest.ColorSpaceTest) ... ok
test_bitdepth (ColorSpaceTest.ColorSpaceTest) ... ok
test_category (ColorSpaceTest.ColorSpaceTest) ... ok
test_config (ColorSpaceTest.ColorSpaceTest) ... ERROR
test_constructor_with_keyword (ColorSpaceTest.ColorSpaceTest) ... ok
test_constructor_without_keyword (ColorSpaceTest.ColorSpaceTest) ... ok
test_constructor_without_parameter (ColorSpaceTest.ColorSpaceTest) ... ok
test_data (ColorSpaceTest.ColorSpaceTest) ... ok
test_description (ColorSpaceTest.ColorSpaceTest) ... ok
test_equality (ColorSpaceTest.ColorSpaceTest) ... ok
test_family (ColorSpaceTest.ColorSpaceTest) ... ok
test_name (ColorSpaceTest.ColorSpaceTest) ... ok
test_transform (ColorSpaceTest.ColorSpaceTest) ... ok
test_constructor_with_keyword (CDLTransformTest.CDLTransformTest) ... FAIL
test_constructor_without_keyword (CDLTransformTest.CDLTransformTest) ... FAIL
test_constructor_without_parameter (CDLTransformTest.CDLTransformTest) ... FAIL
test_createfromfile_cc (CDLTransformTest.CDLTransformTest) ... ok
test_createfromfile_ccc (CDLTransformTest.CDLTransformTest) ... ok
test_createfromfile_cdl (CDLTransformTest.CDLTransformTest) ... [OpenColorIO Warning]: File name not specified(16): Unrecognized element 'MediaRef' where its parent is 'ColorDecision' (11): : Unknown element.
[OpenColorIO Warning]: File name not specified(42): Unrecognized element 'MediaRef' where its parent is 'ColorDecision' (37): : Unknown element.
ERROR
test_description (CDLTransformTest.CDLTransformTest) ... ok
test_direction (CDLTransformTest.CDLTransformTest) ... FAIL
test_equality (CDLTransformTest.CDLTransformTest) ... ok
test_id (CDLTransformTest.CDLTransformTest) ... ok
test_offset (CDLTransformTest.CDLTransformTest) ... ok
test_power (CDLTransformTest.CDLTransformTest) ... ok
test_saturation (CDLTransformTest.CDLTransformTest) ... ok
test_slope (CDLTransformTest.CDLTransformTest) ... ok
test_style (CDLTransformTest.CDLTransformTest) ... ok
test_validate_direction (CDLTransformTest.CDLTransformTest) ... FAIL
test_validate_power (CDLTransformTest.CDLTransformTest) ... ok
test_validate_saturation (CDLTransformTest.CDLTransformTest) ... ok
test_validate_slope (CDLTransformTest.CDLTransformTest) ... ok
test_xml (CDLTransformTest.CDLTransformTest) ... 
<ColorCorrection id="shot 042">
    <SOPNode>
        <Description>Cool look for forest scenes.</Description>
        <Slope>1.5 2 2.5</Slope>
        <Offset>3 3.5 4</Offset>
        <Power>4.5 5 5.5</Power>
    </SOPNode>
    <SatNode>
        <Saturation>6</Saturation>
    </SatNode>
</ColorCorrection>

FAIL

======================================================================
ERROR: test_config (ColorSpaceTest.ColorSpaceTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/meichu/CodeProjects/OpenColorIO/tests/python/ColorSpaceTest.py", line 147, in test_config
    transforms = to_ref.getTransforms()
AttributeError: 'PyOpenColorIO.GroupTransform' object has no attribute 'getTransforms'

======================================================================
ERROR: test_createfromfile_cdl (CDLTransformTest.CDLTransformTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/meichu/CodeProjects/OpenColorIO/tests/python/CDLTransformTest.py", line 246, in test_createfromfile_cdl
    cdl2 = OCIO.CDLTransform.CreateFromFile(test_file, 'cc0003')
PyOpenColorIO.Exception: The specified cccid/cccindex 'cc0003' could not be loaded from the src file 'tests/data/files/cdl_test1.cdl'.

======================================================================
FAIL: test_constructor_with_keyword (CDLTransformTest.CDLTransformTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/meichu/CodeProjects/OpenColorIO/tests/python/CDLTransformTest.py", line 346, in test_constructor_with_keyword
    self.assertEqual(self.TEST_CDL, cdl_tr.getXML())
AssertionError: '\n<ColorCorrection id="shot 042">\n    <S[275 chars]n>\n' != '<ColorCorrection id="shot 042">\n    <SOP[271 chars]ion>'
- 
  <ColorCorrection id="shot 042">
      <SOPNode>
          <Description>Cool look for forest scenes.</Description>
          <Slope>1.5 2 2.5</Slope>
          <Offset>3 3.5 4</Offset>
          <Power>4.5 5 5.5</Power>
      </SOPNode>
      <SatNode>
          <Saturation>6</Saturation>
      </SatNode>
- </ColorCorrection>
?                   -
+ </ColorCorrection>

======================================================================
FAIL: test_constructor_without_keyword (CDLTransformTest.CDLTransformTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/meichu/CodeProjects/OpenColorIO/tests/python/CDLTransformTest.py", line 382, in test_constructor_without_keyword
    self.assertEqual(self.TEST_CDL, cdl_tr.getXML())
AssertionError: '\n<ColorCorrection id="shot 042">\n    <S[275 chars]n>\n' != '<ColorCorrection id="shot 042">\n    <SOP[271 chars]ion>'
- 
  <ColorCorrection id="shot 042">
      <SOPNode>
          <Description>Cool look for forest scenes.</Description>
          <Slope>1.5 2 2.5</Slope>
          <Offset>3 3.5 4</Offset>
          <Power>4.5 5 5.5</Power>
      </SOPNode>
      <SatNode>
          <Saturation>6</Saturation>
      </SatNode>
- </ColorCorrection>
?                   -
+ </ColorCorrection>

======================================================================
FAIL: test_constructor_without_parameter (CDLTransformTest.CDLTransformTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/meichu/CodeProjects/OpenColorIO/tests/python/CDLTransformTest.py", line 397, in test_constructor_without_parameter
    self.assertEqual(cdl_tr.getXML(), DEFAULT_CDL)
AssertionError: '<ColorCorrection id="">\n    <SOPNode>\n [188 chars]ion>' != '\n<ColorCorrection id="">\n    <SOPNode>\[192 chars]n>\n'
+ 
  <ColorCorrection id="">
      <SOPNode>
          <Slope>1 1 1</Slope>
          <Offset>0 0 0</Offset>
          <Power>1 1 1</Power>
      </SOPNode>
      <SatNode>
          <Saturation>1</Saturation>
      </SatNode>
- </ColorCorrection>+ </ColorCorrection>
?                   +


======================================================================
FAIL: test_direction (CDLTransformTest.CDLTransformTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/meichu/CodeProjects/OpenColorIO/tests/python/CDLTransformTest.py", line 163, in test_direction
    self.assertEqual(self.cdl_tr.getDirection(), direction)
AssertionError: TransformDirection.TRANSFORM_DIR_FORWARD != TransformDirection.TRANSFORM_DIR_UNKNOWN

======================================================================
FAIL: test_validate_direction (CDLTransformTest.CDLTransformTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/meichu/CodeProjects/OpenColorIO/tests/python/CDLTransformTest.py", line 305, in test_validate_direction
    self.cdl_tr.validate()
AssertionError: Exception not raised

======================================================================
FAIL: test_xml (CDLTransformTest.CDLTransformTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/meichu/CodeProjects/OpenColorIO/tests/python/CDLTransformTest.py", line 193, in test_xml
    self.assertEqual(self.cdl_tr.getXML(), self.TEST_CDL)
AssertionError: '<ColorCorrection id="shot 042">\n    <SOP[271 chars]ion>' != '\n<ColorCorrection id="shot 042">\n    <S[275 chars]n>\n'
+ 
  <ColorCorrection id="shot 042">
      <SOPNode>
          <Description>Cool look for forest scenes.</Description>
          <Slope>1.5 2 2.5</Slope>
          <Offset>3 3.5 4</Offset>
          <Power>4.5 5 5.5</Power>
      </SOPNode>
      <SatNode>
          <Saturation>6</Saturation>
      </SatNode>
- </ColorCorrection>+ </ColorCorrection>
?                   +


----------------------------------------------------------------------
Ran 34 tests in 0.016s

FAILED (failures=6, errors=2)
```

**Comment by doug-walker** [comment]
This is super helpful, thanks for this contribution @meimchu!   Patrick, Bernard, and I will give it a careful review this week.

**Comment by hodoulp** [comment]
Hi @meimchu,
Thanks a lot for the time you put in that important effort.

To help you, the included file is a patch file to fix all the issues around the changed Python unit tests. As you are currently rewritten them, the untouched ones were only partially fixed (with the most obvious changes to highlight the differences) and disabled for now. But if you apply the patch as-is, the build & tests will succeed.

[python_unit_tests_patch.txt](https://github.com/AcademySoftwareFoundation/OpenColorIO/files/4659088/python_unit_tests_patch.txt)

**Comment by meimchu** [comment]
This patch looks great. I will play around with it and apply the patch work for the UnitTestUtils, OpenColorIOTestSuite, CDLTransformTest and ColorSpaceTest units. I am looking at picking another module for rewrite next as well.

**Comment by hodoulp** [comment]
As it could take some time to rewrite & improve all the existing unit tests I suggest to apply the patch so this PR (or a new PR containing all the changes) can be reviewed & merged and in parallel you can continue the work.  So, `master` contains some Python binding unit tests up and running and the community can contribute independently from your effort (e.g. Autodesk can add Python bindings for new classes, new methods, etc) without any overlaps.


> Should I wait for the FIXME to be rectified from CDLTransform first?

No need to wait. A pull request is coming...

**Comment by meimchu** [comment]
I have applied the patch and re-enabled the specific cdl test given that PR is available. I see this test is failing as expected on CI now that those specific python unit tests are enabled. Once that PR is merged, I am sure this won't trigger a fail anymore. I have rebased this branch now to this latest master.

**Comment by hodoulp** [comment]
This has been up for two weeks, given that there has been no new feedback, we will go ahead and merge next week.

File: tests/python/CDLTransformTest.py [hunk 1]
```diff
@@ -1,109 +1,393 @@
...
+        <Slope>{2}</Slope>
+        <Offset>{3}</Offset>
+        <Power>{4}</Power>
+    </SOPNode>
+    <SatNode>
+        <Saturation>{5}</Saturation>
+    </SatNode>
+</ColorCorrection>""".format(TEST_CDL_ID,
+           TEST_CDL_DESC,
+           ' '.join([str(s) for s in TEST_CDL_SLOPE]),
```
[the remark this answers is not in the document]
Comment (meimchu) [reply tests/python/CDLTransformTest.py]: Commit [28a1f63] addressed this.

File: tests/python/ColorSpaceTest.py [hunk 2]
```diff
@@ -1,31 +1,308 @@
...
+    def test_allocation_vars(self):
+        """
+        Test the setAllocationVars() and getAllocationVars() methods.
+        """
+
+        # Array length tests
+        alloc_vars = []
+        for i in range(1, 5):
+            # This will create [0.1] up to [0.1, 0.2, 0.3, 0.4]
+            alloc_vars.append(float('0.%i' % i))
```
[the remark this answers is not in the document]
Comment (meimchu) [reply tests/python/ColorSpaceTest.py]: Commit 28a1f63 addressed this.

File: tests/python/ColorSpaceTest.py [hunk 3]
```diff
@@ -1,31 +1,308 @@
...
+        for wrong_is_data in wrong_is_datas:
+            with self.assertRaises(TypeError):
+                self.colorspace.setIsData(wrong_is_data)
+
+    def test_description(self):
+        """
+        Test the setDescription() and getDescription() methods.
+        """
+
+        # Test with TEST_DESCS
```
[the remark this answers is not in the document]
Comment (meimchu) [reply tests/python/ColorSpaceTest.py]: Commit 28a1f63 addressed this.

**Comment by meimchu** [comment]
I will let this PR merge and then start a new PR for future python unit test updates. I am looking at starting work on LookTest module next.

**Comment by meimchu** [comment]
Replaced with #1025

**Pull request closed without merging**